curta/src/chip/hash/sha/keccak256/mod.rs

```diff
@@ -0,0 +1,166 @@
+use core::borrow::Borrow;
+
+use plonky2::field::types::Field;
+
+use crate::chip::{builder::{AirBuilder}, AirParameters, trace::writer::TraceWriter, uint::{register::{U64Register, ByteArrayRegister}, bytes::lookup_table::builder_operations::ByteLookupOperations, operations::instruction::U32Instructions}, arithmetic::expression, register::bit::BitRegister};
+use crate::chip::register::{Register, RegisterSerializable, RegisterSized};
+
+use super::sha256::U32Value;
+
+pub struct Keccak256Gadget {
+   
+}
+
+#[derive(Debug, Clone)]
+pub struct Keccak256PublicData<T> {
+    pub public_w: Vec<U32Value<T>>,
+    pub hash_state: Vec<U32Value<T>>,
+    pub end_bits: Vec<T>,
+}
+
+#[rustfmt::skip]
+pub const KECCAKF_ROTC: [[u32; 5]; 5] = [
+    [0, 1, 62, 28, 27],
+    [36, 44, 6, 55, 20],
+    [3, 10, 43, 25, 39],
+    [41, 45, 15, 21, 8],
+    [18, 2, 61, 56, 14]
+];
+
+#[rustfmt::skip]
+pub const KECCAKF_RNDC: [[u32; 2]; 24] = [
+    [0x00000001, 0x00000000], [0x00008082, 0x00000000],
+    [0x0000808A, 0x80000000], [0x80008000, 0x80000000],
+    [0x0000808B, 0x00000000], [0x80000001, 0x00000000],
+    [0x80008081, 0x80000000], [0x00008009, 0x80000000],
+    [0x0000008A, 0x00000000], [0x00000088, 0x00000000],
+    [0x80008009, 0x00000000], [0x8000000A, 0x00000000],
+    [0x8000808B, 0x00000000], [0x0000008B, 0x80000000],
+    [0x00008089, 0x80000000], [0x00008003, 0x80000000],
+    [0x00008002, 0x80000000], [0x00000080, 0x80000000],
+    [0x0000800A, 0x00000000], [0x8000000A, 0x80000000],
+    [0x80008081, 0x80000000], [0x00008080, 0x80000000],
+    [0x80000001, 0x00000000], [0x80008008, 0x80000000],
+];
+
+// impl constraint for keccack256 
+// state register 
+impl<L: AirParameters> AirBuilder<L> {
+    pub fn keccak_f(&mut self,
+        operations: &mut ByteLookupOperations,
+    )where
+    // SHOULD it be u64instructions?
+    L::Instruction: U32Instructions {
+        // alloate 5*5 u64 register for storing 25 8byte lanes
+        // state is 25 * 64 bit
```

Comment:
This looks really good so far! I'm really impressed.

Here are some notes based on this reference: https://keccak.team/keccak_specs_summary.html#:~:text=Pseudo%2Dcode%20description%20of%20the%20permutations.

- The high level strategy, from my understanding, is to compute one round of Keccak in a single row in the STARK.
- You need to have a way to keep track of "i", the round index, to know which round constant to use.
- Thus, you should allocate 24 columns and constrain them such that they cycle from:

[1, 0, 0, ..., 0]
[0, 1, 0, ..., 0]
[0, 0, 1, ..., 0]
until it wraps around...

Let's call this set of 24 columns ROUND_CONSTANT_SELECTOR.

- Now, to get the right round constant you should essentially compute a dot product between ROUND_CONSTANT_SELECTOR and ROUND_CONSTANTS which you can do by in a single constraint by allocating another column called SELECTED_ROUND_CONSTANT.

- From that point on, your steps look mostly correct!

Reply:
Another option is to use a Bus like we do in SHA256, but since there are only 24 round constants John's suggestion might be similar in performance. 

For now, I would not worry about constraining the round constants, write them directly to the trace and we can add these constraints in the end.